flink-tests/src/test/java/org/apache/flink/test/checkpointing/ChangelogPeriodicMaterializationSwitchStateBackendITCase.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.test.checkpointing;
+
+import org.apache.flink.runtime.state.AbstractStateBackend;
+import org.apache.flink.streaming.api.environment.CheckpointConfig;
+import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
+
+import org.junit.Test;
+
+/**
+ * This verifies that switching state backend works correctly for Changelog state backend with
+ * materialized state / non-materialized state.
+ */
+public class ChangelogPeriodicMaterializationSwitchStateBackendITCase
```

Comment:
I think the test doesn't cover reading the changelog for PQ states currently, right?
I.e. there are no timers/windows in the graph used.

Reply:
Yeah, I have added a timer in the test base.
I have modified the `ChangelogMigrationRestoreTarget#getRestoredKeyedStateBackend` because pq states do not supports migration currently.
BTW, I also enable switching case in `ChangelogCompatibilityITCase` which uses windows in the graph.

Reply:
> I have added a timer in the test base.

I think the timers are saved but the correctness of restored timers is not validated, is it?
With "normal" window operator, incorrectly restored timer might lead to window not firing or just an exception.

> I have modified the ChangelogMigrationRestoreTarget#getRestoredKeyedStateBackend because pq states do not supports migration currently.

Could you explain what do you mean, why pq states don't support migration?


> BTW, I also enable switching case in ChangelogCompatibilityITCase which uses windows in the graph.

Good idea! :+1:

Reply:
> Could you explain what do you mean, why pq states don't support migration?
I mean "state migration".
beacuse we do not support state migration currently. So while switching from enabled to disabled, the 1st user access will fail in original implementation. You could see TODO in the ChangelogMigrationRestoreTarget#getRestoredKeyedStateBackend.